communications/views.py

```diff
@@ -1,12 +1,148 @@
+        return Response({'unread_count': count})
+    
+    @action(detail=True, methods=['post'])
+    def mark_as_read(self, request, notification_id=None):
+        """Mark a notification as read."""
+        notification = self.get_object()
+        notification.is_read = True
+        notification.save()
+        return Response({'status': 'notification marked as read'})
+    
+    @action(detail=False, methods=['post'])
+    def mark_all_as_read(self, request):
+        """Mark all notifications as read for the current user."""
+        updated = self.get_queryset().filter(is_read=False).update(is_read=True)
+        return Response({'status': f'marked {updated} notifications as read'})
+
+
+class NotificationTypeViewSet(viewsets.ReadOnlyModelViewSet):
+    """
+    API endpoint that allows notification types to be viewed.
+    Requires authentication.
+    """
+    queryset = NotificationType.objects.filter(is_active=True)
+    serializer_class = NotificationTypeSerializer
+    permission_classes = [permissions.IsAuthenticated]
+    pagination_class = None
+    
+    def get_queryset(self):
+        """Return only active notification types."""
+        return NotificationType.objects.filter(is_active=True)
+
+
+
+class UserNotificationPreferenceViewSet(viewsets.ModelViewSet):
+    """
+    API endpoint that allows users to view and update their notification preferences.
+    """
+    serializer_class = UserNotificationPreferenceSerializer
+    permission_classes = [permissions.IsAuthenticated]
+    http_method_names = ['get', 'patch', 'head', 'options']
+
+    def get_queryset(self):
+        """Return only the current user's preferences."""
+        return UserNotificationPreference.objects.filter(user=self.request.user)
+
+    def get_object(self):
+        """
+        Return the current user's preferences, creating them if they don't exist.
+        """
+        queryset = self.filter_queryset(self.get_queryset())
+        obj = queryset.first()
+        
+        if obj is None:
+            obj = UserNotificationPreference.objects.create(user=self.request.user)
+            
+            notification_types = NotificationType.objects.filter(is_active=True)
+            for notification_type in notification_types:
+                UserNotificationTypePreference.objects.create(
+                    user_preference=obj,
+                    notification_type=notification_type,
+                    frequency='immediate'
+                )
+        
+        return obj
+
+    @action(detail=True, methods=['patch'])
+    def update_type_preference(self, request, pk=None):
+        """
+        Update a specific notification type preference.
+        Expected payload: {"notification_type_id": 1, "frequency": "immediate"}
+        """
+        preference = self.get_object()
+        notification_type_id = request.data.get('notification_type_id')
+        frequency = request.data.get('frequency')
```

Comment:
frequency is not validated against allowed choices. Before saving, check that the value is in NotificationFrequency.choices. Otherwise, it could result in an invalid state or error.